src/main/java/seedu/task/logic/commands/DeleteTaskCommand.java

```diff
@@ -12,30 +13,60 @@
 
     public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
 
+    private ReadOnlyTask taskToDelete;
+    
     public DeleteTaskCommand(int targetIndex) {
-        this.targetIndex = targetIndex;
+        this.lastShownListIndex = targetIndex;
     }
 
 
-    @Override
+    public DeleteTaskCommand(ReadOnlyTask taskToDelete) {
+		this.taskToDelete = taskToDelete;
+	}
+
+
+	@Override
     public CommandResult execute() {
 
         UnmodifiableObservableList<ReadOnlyTask> lastShownList = model.getFilteredTaskList();
 
-        if (lastShownList.size() < targetIndex) {
+        if (lastShownList.size() < lastShownListIndex) {
             indicateAttemptToExecuteIncorrectCommand();
             return new CommandResult(Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
         }
-
-        ReadOnlyTask taskToDelete = lastShownList.get(targetIndex - 1);
-
+        
+        if(lastShownListIndex != 0) {
+        	taskToDelete = lastShownList.get(lastShownListIndex - 1);
+        }
+        
         try {
             model.deleteTask(taskToDelete);
         } catch (TaskNotFoundException tnfe) {
             assert false : "The target task cannot be missing";
         }
-
+        
+        reverseCommand = prepareUndoCommand();
         return new CommandResult(String.format(MESSAGE_DELETE_TASK_SUCCESS, taskToDelete));
     }
 
+
+	@Override
+	public UndoableCommand prepareUndoCommand() {
+		UndoableCommand command = new AddTaskCommand(taskToDelete); 
+		command.setData(model);
+		
```

Comment:
it is to link the command to the model. 
Otherwise when the command is executed later when the user undoes, it will have no way to modify the model.  It is different from the other implementation which saves a snapshot of model because it only saves a dependency.